grove/tests/measurements/test_estimation.py

```diff
@@ -0,0 +1,148 @@
+"""
+Tests for the estimation module
+"""
+import pytest
+from mock import Mock
+import numpy as np
+from scipy.stats import bernoulli
+from pyquil.paulis import sX, sY, sZ, sI, PauliSum, is_zero
+from pyquil.quil import Program
+from pyquil.gates import RY, RX
+from pyquil.api import QVMConnection
+from grove.measurements.estimation import (remove_imaginary_terms,
+                                           get_rotation_program,
+                                           get_parity,
+                                           estimate_pauli_set,
+                                           CommutationError)
+
+
+def test_imaginary_removal():
+    """
+    remove terms with imaginary coefficients from a pauli sum
+    """
+    test_term = 0.25 * sX(1) * sZ(2) * sX(3) + 0.25j * sX(1) * sZ(2) * sY(3)
+    test_term += -0.25j * sY(1) * sZ(2) * sX(3) + 0.25 * sY(1) * sZ(2) * sY(3)
+    true_term = 0.25 * sX(1) * sZ(2) * sX(3) + 0.25 * sY(1) * sZ(2) * sY(3)
+    assert remove_imaginary_terms(test_term) == true_term
+
+    test_term = (0.25 + 1j) * sX(0) * sZ(2) + 1j * sZ(2)
+    # is_identity in pyquil apparently thinks zero is identity
+    assert is_zero(remove_imaginary_terms(test_term))
+
+    test_term = 0.25 * sX(0) * sZ(2) + 1j * sZ(2)
+    assert remove_imaginary_terms(test_term) == PauliSum([0.25 * sX(0) * sZ(2)])
+
+    with pytest.raises(TypeError):
+        remove_imaginary_terms(5)
+
+    with pytest.raises(TypeError):
+        remove_imaginary_terms(sX(0))
+
+
+def test_rotation_programs():
+    """
+    Testing the generation of post rotations
+    """
+    test_term = sZ(0) * sX(20) * sI(100) * sY(5)
+    # note: string comparison of programs requires gates to be in the same order
+    true_rotation_program = Program().inst(
+        [RX(np.pi / 2)(5), RY(-np.pi / 2)(20)])
+    test_rotation_program = get_rotation_program(test_term)
+    assert true_rotation_program.out() == test_rotation_program.out()
+
+
+def test_get_parity():
+    """
+    Check if our way to compute parity is correct
+    """
+    single_qubit_results = [[0]] * 50 + [[1]] * 50
+    single_qubit_parity_results = list(map(lambda x: -2 * x[0] + 1,
+                                           single_qubit_results))
+
+    # just making sure I constructed my test properly
+    assert np.allclose(np.array([1] * 50 + [-1] * 50),
+                       single_qubit_parity_results)
+
+    test_results = get_parity([sZ(5)], single_qubit_results)
+    assert np.allclose(single_qubit_parity_results, test_results[0, :])
+
+    np.random.seed(87655678)
+    brv1 = bernoulli(p=0.25)
+    brv2 = bernoulli(p=0.4)
+    n = 500
+    two_qubit_measurements = list(zip(brv1.rvs(size=n), brv2.rvs(size=n)))
+    # note: get_parity doesn't check if basis makes sense. The following example
+    # would not work on the device
```

Comment:
You mean because sX(1) and sZ(0)*sZ(1) do not commute? Why is it necessary to use this in the test?

Reply:
yes.  The Pauli terms in this test are used to indicate which position to look at in the measurement result bit string.  I will change this to a working example and delete the note.